app/updatemanager.go

```diff
@@ -0,0 +1,155 @@
+
+func init() {
+	UpdateManagerMap = NewControlMap()
+}
+
+type UpdateManager struct {
+	dbus                        dbus.DBusAPI
+	dbusConn                    dbus.Handle
+	updateControlTimeoutSeconds int
+}
+
+func NewUpdateManager(updateControlTimeoutSeconds int) *UpdateManager {
+	return &UpdateManager{
+		updateControlTimeoutSeconds: updateControlTimeoutSeconds,
+	}
+	return &UpdateManager{}
+}
+
+func (u *UpdateManager) EnableDBus(api dbus.DBusAPI) {
+	u.dbus = api
+}
+
+func (u *UpdateManager) Start() (context.CancelFunc, error) {
+	log.Debug("Running the UpdateManager")
+	if u.dbus == nil {
+		return nil, errors.New("DBus not enabled")
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	go u.run(ctx)
+	return cancel, nil
+}
+
+func (u *UpdateManager) run(ctx context.Context) error {
+	log.Debug("Running the UpdateManager")
+	dbusConn, err := u.dbus.BusGet(dbus.GBusTypeSystem)
+	if err != nil {
+		return errors.Wrap(err, "Failed to start the Update manager")
+	}
+	nameGID, err := u.dbus.BusOwnNameOnConnection(
+		dbusConn,
+		UpdateManagerDBusObjectName,
+		dbus.DBusNameOwnerFlagsAllowReplacement|dbus.DBusNameOwnerFlagsReplace,
+	)
+	if err != nil {
+		return errors.Wrap(err, "Failed to start the update manager")
+	}
+	defer u.dbus.BusUnownName(nameGID)
+	intGID, err := u.dbus.BusRegisterInterface(
+		dbusConn,
+		UpdateManagerDBusPath,
+		UpdateManagerDBusInterface,
+	)
+	if err != nil {
+		log.Errorf("Failed to register the DBus interface name %q at path %q: %s",
+			UpdateManagerDBusInterface,
+			UpdateManagerDBusPath,
+			err,
+		)
+		return err
+	}
+	defer u.dbus.BusUnregisterInterface(dbusConn, intGID)
+
+	u.dbus.RegisterMethodCallCallback(
+		UpdateManagerDBusPath,
+		UpdateManagerDBusInterfaceName,
+		"SetUpdateControlMap",
+		func(objectPath string, interfaceName string, methodName string, updateControlMap string) (interface{}, error) {
+			controlMap := UpdateControlMap{}
+			err := json.Unmarshal([]byte(updateControlMap), &controlMap)
+			if err != nil {
+				log.Errorf("Failed to unmarshal the JSON in the string received via SetControlMap on D-Bus: %s", err)
+				return u.updateControlTimeoutSeconds / 2, err
+			}
+			log.Debugf("Setting the control map parameter to: %s", controlMap.ID)
+			UpdateManagerMap.Set(&controlMap)
+			return u.updateControlTimeoutSeconds / 2, nil
+		})
+	defer u.dbus.UnregisterMethodCallCallback(
+		UpdateManagerDBusPath,
+		UpdateManagerDBusInterfaceName,
+		"SetUpdateControlMap")
+	<-ctx.Done()
+  return nil
+}
+
+type UpdateControlMap struct {
+	ID       string `json:"id"`
+	Priority int    `json:"priority"`
+}
+
+type ControlMap struct {
+	controlMap map[string][]*UpdateControlMap
+	mutex      sync.Mutex
+}
+
+func NewControlMap() *ControlMap {
+	return &ControlMap{
+		controlMap: make(map[string][]*UpdateControlMap),
+	}
+}
+
+func (c *ControlMap) Set(cm *UpdateControlMap) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	c.controlMap[cm.ID] = append(c.controlMap[cm.ID], cm)
```

Comment:
Maybe you misunderstood my original comment. When storing the map, all pairs of ID and priority must be unique. The way you stored it earlier did do this correctly (though lookup was tricky). However, here you have lost the uniqueness: If I update using a map with the same ID and priority, it will be appended to the end of the list instead of being replaced.

Here's my suggestion for nailing this in a test: Use my exact example from [MEN-4535](https://tracker.mender.io/browse/MEN-4535). For the first three cases, verify that inserting the two maps returns two maps as well. For the last case, verify that it returns only one map, the last one.

What data structure to use, hmm, you could use your original one containing both ID and priotity. You'd have to do a linear search to look up by ID, but this is fine, we are not expecting many maps, so performance is not a concern.